ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound/Registry.hs

```diff
@@ -210,8 +221,43 @@ withPeer tracer
                       -> Map txid tx
                       -- ^ received txs
                       -> m ()
-    handleReceivedTxs txids txs =
+    handleReceivedTxs txids txs = do
+      void $ atomically $ modifyTVar sharedStateVar addFethed
       collectTxs tracer sharedStateVar peeraddr txids txs
+       where
+        addFethed :: SharedTxState peeraddr txid tx
+                  -> SharedTxState peeraddr txid tx
+        addFethed st@SharedTxState { peerTxStates } =
+          let peerTxStates' = Map.update (\ps -> Just $! ps { fetchedTxs = Set.union (fetchedTxs ps) txids }) peeraddr peerTxStates in
+          st {peerTxStates = peerTxStates' }
+
+    countRejectedTxs :: Int
+                     -> m Int
+    countRejectedTxs n = atomically $ do
+      modifyTVar sharedStateVar cntRejects
+      st <- readTVar sharedStateVar
+      case Map.lookup peeraddr (peerTxStates st)  of
+           Nothing -> error "missing peer updated"
+           Just ps -> return $ rejectedTxs ps
+       where
+        cntRejects :: SharedTxState peeraddr txid tx
+                   -> SharedTxState peeraddr txid tx
+        cntRejects st@SharedTxState { peerTxStates } =
+          let peerTxStates' = Map.update (\ps -> Just $! ps { rejectedTxs = min 42 (max (-42) (rejectedTxs ps + n)) }) peeraddr peerTxStates in
+          st {peerTxStates    = peerTxStates'}
```

Comment:
It will be more efficient to update it for all peers at once, but it might be harder to implement it. Another way would be to store the counts in `Map peerAddt (TVar ...)`, so that we don't block on taking the whole state to update the count just for one peer.